bionic/deriver.py

```diff
@@ -340,7 +340,29 @@ def _compute_result_group_for_dnode(self, dnode):
             self._initialize_task_state(state)
 
             # If the task isn't complete or blocked, we can complete the task.
-            complete_task_state(state, task_key_logger)
+
+            # self._executor is None when parallel processing is disabled or when
+            # entities are bootstrapped.
+            if not state.provider.attrs.should_persist() or self._executor is None:
+                # NOTE 1: This makes non-persisted entities compute here as well
+                # as in executor pool. We need to keep track of what needs to be
+                # computed in main process vs subprocess when entity is not persisted.
+                # NOTE 2: Right now, non-persisted entities include simple lookup values
+                # which we should not be really sending using IPC. We should read/write
+                # a tmp file for this instead to use protocol for serialization instead of
+                # using cloudpickle.
+                complete_task_state(state, task_key_logger)
+            else:
+                # NOTE 1: Logging support for multiple processes not done yet.
+                # NOTE 2: Non-persisted entities are not computed inside the executor yet.
+                # Right now the tests pass because all non-persisted entities are serializable
+                # but that won't always be the case.
+                ex = self._executor.submit(complete_task_state, state, task_key_logger)
+                ex.result()
+                state.is_complete = True
```

Comment:
Maybe we should set this at the end of `_populate_task_state_value_hash`? Strictly speaking, this shouldn't be true until `result_value_hashes_by_name` is set, right? (For example, suppose we threw an exception in between setting `is_complete` and setting `result_value_hashes_by_name`)?

Reply:
Yeah, that's fair. I'll move it after populating value hashes.